File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/e1e68be6e0ed98370439e1e660ed4c97.c
#include <stdio.h>

#define MOD 998244353

typedef long long LL;

LL factorial[3123456];
LL inverse_factorial[3123456];

LL add(LL a, LL b) {
    return (a + b) % MOD;
}

LL mul(LL a, LL b) {
    return (a * b) % MOD;
}

LL pow(LL a, LL x) {
    LL ret = 1;
    while (x) {
        if (x & 1) {
            ret = mul(ret, a);
        }
        a = mul(a, a);
        x >>= 1;
    }
    return ret;
}

LL inv(LL x) {
    LL a = x, b = MOD, u = 1, v = 0;
    while (b) {
        LL t = a / b;
        a -= t * b;
        u -= t * v;
        LL tmp = a;
        a = b;
        b = tmp;
        tmp = u;
        u = v;
        v = tmp;
    }
    return (u + MOD) % MOD;
}

void build_factorial() {
    factorial[0] = 1;
    for (int i = 1; i < 3123456; i++) {
        factorial[i] = mul(factorial[i - 1], i);
    }
    inverse_factorial[3123456 - 1] = inv(factorial[3123456 - 1]);
    for (int i = 3123456 - 1; i >= 1; i--) {
        inverse_factorial[i - 1] = mul(inverse_factorial[i], i);
    }
}

LL combination(int n, int k) {
    if (k < 0 || k > n) return 0;
    return mul(mul(factorial[n], inverse_factorial[k]), inverse_factorial[n - k]);
}

LL multi_choose(int n, int r) {
    if (n == 0) {
        return (r == 0 ? 1 : 0);
    }
    if (r == 0) {
        return 1;
    }
    return combination(n + r - 1, r);
}

LL multi_choose_with_limit(int n, int r, int lim) {
    if (n == 0) {
        return (r == 0 ? 1 : 0);
    }
    if (lim <= 0 && r > 0) {
        return 0;
    }
    if (r == 0) {
        return (lim < 0 ? 0 : 1);
    }
    LL ret = 0;
    for (int i = 0; i <= n; i++) {
        if (i * (lim + 1) > r) break;
        ret += ((i & 1) ? MOD - 1 : 1) * combination(n, i) * multi_choose(n, r - i * (lim + 1));
        ret %= MOD;
    }
    return ret;
}

int main() {
    int p, s, r;
    scanf("%d%d%d", &p, &s, &r);
    build_factorial();
    LL ret = 0;
    for (int top = r; top <= s; top++) {
        LL tot = 0;
        for (int q = 0; q < p; q++) {
            if (top * q > s - top) break;
            tot += mul(mul(combination(p - 1, q), multi_choose_with_limit(p - q - 1, s - top - top * q, top - 1)), inv(q + 1));
            tot %= MOD;
        }
        ret += tot;
        ret %= MOD;
    }
    ret = mul(ret, inv(multi_choose(p, s - r)));
    printf("%lld\n", ret);
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/e1e68be6e0ed98370439e1e660ed4c97.java
import java.util.*;
import java.io.*;

class ModInt {
    private int x;
    private static int mod;

    public ModInt() {
        x = 0;
    }

    public ModInt(int y) {
        x = y;
    }

    public ModInt(long y) {
        x = (int) ((mod + y % mod) % mod);
    }

    public static void setMod(int m) {
        mod = m;
    }

    public long value() {
        return x;
    }

    public ModInt add(ModInt r) {
        long sum = this.value() + r.value();
        return new ModInt(sum >= mod ? sum - mod : sum);
    }

    public ModInt multiply(ModInt r) {
        return new ModInt((this.value() * r.value()) % mod);
    }

    public ModInt pow(long exp) {
        ModInt ret = new ModInt(1);
        ModInt base = this;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                ret = ret.multiply(base);
            }
            base = base.multiply(base);
            exp >>= 1;
        }
        return ret;
    }

    public ModInt inv() {
        long a = this.value(), b = mod, u = 1, v = 0;
        while (b > 0) {
            long t = a / b;
            long temp = a; a = b; b = temp - t * b;
            temp = u; u = v; v = temp - t * v;
        }
        return new ModInt(u);
    }

    public ModInt subtract(ModInt r) {
        return this.add(new ModInt(mod - r.value()));
    }

    public ModInt divide(ModInt r) {
        return this.multiply(r.inv());
    }

    public static ModInt combination(int n, int k) {
        if (k < 0 || k > n) return new ModInt(0);
        return factorial[n].multiply(inverseFactorial[k]).multiply(inverseFactorial[n - k]);
    }

    private static ModInt[] factorial = new ModInt[3123456];
    private static ModInt[] inverseFactorial = new ModInt[3123456];
    private static boolean isBuilt = false;

    public static void build() {
        isBuilt = true;
        factorial[0] = new ModInt(1);
        for (int i = 1; i < factorial.length; i++) {
            factorial[i] = factorial[i - 1].multiply(new ModInt(i));
        }
        inverseFactorial[factorial.length - 1] = factorial[factorial.length - 1].inv();
        for (int i = factorial.length - 1; i >= 1; i--) {
            inverseFactorial[i - 1] = inverseFactorial[i].multiply(new ModInt(i));
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int p = Integer.parseInt(st.nextToken());
        int s = Integer.parseInt(st.nextToken());
        int r = Integer.parseInt(st.nextToken());

        ModInt.setMod(1000000007); // Set your mod value here
        if (!isBuilt) build();

        ModInt ret = new ModInt(0);
        for (int top = r; top <= s; top++) {
            ModInt tot = new ModInt(0);
            for (int q = 0; q < p; q++) {
                if (top * q > s - top) break;
                tot = tot.add(factorial[p - 1].divide(factorial[q]).multiply(multiChoose(p - q - 1, s - top - top * q, top - 1)).divide(new ModInt(q + 1)));
            }
            ret = ret.add(tot);
        }
        ret = ret.multiply(inv(multiChoose(p, s - r)));
        System.out.println(ret.value());
    }

    public static ModInt multiChoose(int n, int r, int lim) {
        if (n == 0) {
            return new ModInt(r == 0 ? 1 : 0);
        }
        if (lim <= 0 && r > 0) {
            return new ModInt(0);
        }
        if (r == 0) {
            return new ModInt(lim < 0 ? 0 : 1);
        }
        ModInt ret = new ModInt(0);
        for (int i = 0; i <= n; i++) {
            if (i * (lim + 1) > r) break;
            ret = ret.add(((i & 1) == 1 ? mod - 1 : 1) * combination(n, i).multiply(multiChoose(n, r - i * (lim + 1)));
        }
        return ret;
    }
}
